daliuge-engine/dlg/event.py

```diff
@@ -106,10 +117,14 @@ def _fireEvent(self, eventType, **attrs):
 
         # Now that we are sure there are listeners for our event
         # create it and send it to all of them
-        e = Event()
-        e.type = eventType
+        e = Event(eventType)
+
         for k, v in attrs.items():
             setattr(e, k, v)
 
         for l in listeners:
-            l.handleEvent(e)
+            try:
+                l.handleEvent(e)
+            except:
+                logger.exception("Exception in listener %s while handling event %s", l, e)
+                raise
```

Comment:
I suppose this is the key point of your comment: should we keep the existing, sub-optimal behavior or remove this `raise` in favour of having better behavior. If you want to keep this simple let's leave the current behavior, but please create a new ticket (and add a TODO here) to change things in the future. Alternatively if you want to change behavior I'd say let's do it, the unit tests can probably be adjusted/removed if necessary (can you point at them? I'm curious to see whether there are valid reasons for them to exist in the first place, maybe the functionality they test can be tested somehow else).

Also, if choosing to keep the current behavior, I'm not sure this is the best place to place the exception logging code. There are two downsides: first, logging an exception AND re-raising is not usually what you want to do, it leads usually to double-log the exception. But secondly, even though you are logging these particular exceptions, there could be other exceptions that are still *not* being logged in the execution context where this one wasn't originally.

My proposal would be to go to the highest level of execution and try/catch/log exceptions there, at which point you don't need to re-raise them anymore. The threads that run this code are launched from https://github.com/ICRAR/daliuge/blob/adeca035c8a826220e5be2eb2ac15a69f0c4c711/daliuge-engine/dlg/drop.py#L1930. Note that there are two cases here: manually creating a thread and using a thread pool. In both cases you probably want to add a new `def _execute(self...)` method to invoke from `async_execute`, then have the try/self.execute()/catch/log block in `_execute`.

Slightly related: I went to the ticket (https://icrar.atlassian.net/browse/LIU-284) to understand what was the error that triggered this PR but the ticket was unrelated. After some looking I found https://icrar.atlassian.net/browse/LIU-322 that seems to be the real stuff. Is that right?

Reply:
Oh I've totally messed up the ticket number here :facepalm: Yes it is meant to be LIU-322 - I will rename where possible (can't do the branch without recreating the PR)

> My proposal would be to go to the highest level of execution and try/catch/log exceptions there, at which point you don't need to re-raise them anymore. The threads that run this code are launched from

That was my initial thought too, but I figured that this might not actually be the only place where this occurred. Ideally we'd have some sort of global exception handler as exception which is not handled will not be logged. It seems the only place you can do this in Python is in a catch in the `main` entry point. Looking at the stack trace in the ticket it seems we startup this code in a different way - I assume that's to do with the distributed manner of execution here. Is there an entry point on drop execution that I could wrap? I think that would be a much better place to the catch and would handle other cases too. Or is that the `execute` method? (`async_execute` already calls `execute`).

> Alternatively if you want to change behavior I'd say let's do it, the unit tests can probably be adjusted/removed if necessary (can you point at them? I'm curious to see whether there are valid reasons for them to exist in the first place, maybe the functionality they test can be tested somehow else).

I'm happy to change the behaviour here (though there might not be any point if I just catch at the entry point as above). The test that fails is https://github.com/ICRAR/daliuge/blob/adeca035c8a826220e5be2eb2ac15a69f0c4c711/daliuge-engine/test/test_drop.py#L883 where the `n_effective_inputs` assertion fails. 

I think you're right though, catching at the entry point is probably the way to go and will handle any use case. If you're able to confirm `execute()` is the entrypoint then I think I'll do that :)

Reply:
Yes, `async_execute` and `execute` is the place you want to being at. The stack trace Andreas put in LIU-322 starts basically at `threading.run`, signalling this code is running on a different thread than the main one. The place where that thread is created, or at least the execution of `execute()` is scheduled, is in `async_execute`. You can leave the behavior of `execute` untouched by creating a new `_execute` method that invokes `execute` and does the error handling, then schedule `_execute` from `async_execute`.

The test seems sensible, but would indeed conflict with changing the way we handle errors when passing events though the error handles. Let's then leave the behavior of the event handlers as-is and kick that other problem to our future selves.

Reply:
Yep agreed. I will close this PR given the incorrect ticket name and that I'll be redoing this work :upside_down_face: